run-process: fix handling of check-status

phillipwood opened the pull request:
The documentation states that
  If the check-status?: argument is true, then the process exit status
  will be checked, and an error will be raised if it isn't 0.

In fact the argument is named check-status without the question mark
and setting it to '#t' or '#f' results in an error
  *** ERROR IN std/misc/process#run-process__% -- Operator is not a
      PROCEDURE

Fix this by checking if check-status is true rather than just trying
to call it. This potentially breaks existing code that was using the
undocumented behavior and passing a procedure to be called rather than
a flag.

fare commented:
It might be that the error is with the documentation.

Currently, a function is passed that takes two argument, and handles error cases. By passing the default function `check-process-status`, the usual behavior of "throw an error if the exit code wasn't 0" is used. By passing the function `void` (or `true`) the error status can be ignored. By passing a custom function, error status conventions different from the default one can be used.

fare commented:
NB: I am the one who wrote that bad code + documentation.

I lean towards fixing the documentation rather than changing the code. What do you think?

phillipwood commented:
I'm not sure that having `check-status: true` actually ignore the exit status is very intuitive. `check-process-status` isn't currently exported. I'd like to be able to pass `#f` when I want to ignore the status as I think that is clearest. It would also be convenient if `#t` ran the default function rather than having to remember it's name each time. If you want to be able to pass a custom function I can change the patch so that it checks `check-status` with `procedure?` first before checking if it's truthy. In the commit that added this function you talk about it being similar to uiop:run-program; as far as I can see that just uses a flag for checking the exit status.

phillipwood commented:
I've updated the patch so that `check-status:` can take a function as well as `#t` or `#f` and updated the documentation for that and the return value of `run-process`
